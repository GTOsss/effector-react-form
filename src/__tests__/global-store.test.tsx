// TODO need to finish
import React from 'react';
import {render, fireEvent, screen} from '@testing-library/react';
import {createStore} from 'effector';
import {useForm} from '../';
import {OnSubmit} from '../../index';

interface Values {
  username?: string,
  profile?: {
    firstName?: string,
    lastName?: string,
  }
}

const initialState: Values = {username: 'Initial username'};

const $values = createStore(initialState);

const Input = ({
  controller,
  label,
}) => {
  const {input} = controller();

  return (
    <div className="input-wrap">
      <label>{label}</label>
      <input
        {...input}
        value={input.value || ''}
        role="textbox"
        className="input"
        placeholder={label}
      />
    </div>
  );
};

const SimpleForm = () => {
  const {handleSubmit, controller} = useForm<Values>({$values});

  const onSubmit: OnSubmit<Values> = ({values}) => {
    alert(JSON.stringify(values, null, '  '));
  };

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <Input
        label="Username"
        controller={controller({name: 'username'})}
      />
      <Input
        label="First name"
        controller={controller({name: 'profile.firstName'})}
      />
      <Input
        label="Last name"
        controller={controller({name: 'profile.lastName'})}
      />
      <button type="submit">submit</button>
    </form>
  );
}


describe('GlobalStore', () => {
  test('onChange username', () => {
    render(<SimpleForm />);
    const input = screen.getByPlaceholderText('Username');
    fireEvent.change(input, {target: {value: 'test'}});
    const inputs = screen.getAllByRole('textbox');
    expect(inputs).toMatchSnapshot();
  });

  test('onChange username ($values)', () => {
    render(<SimpleForm />);
    const input = screen.getByPlaceholderText('Username');
    fireEvent.change(input, {target: {value: 'test'}});
    expect($values.getState()).toMatchSnapshot();
  });
});


