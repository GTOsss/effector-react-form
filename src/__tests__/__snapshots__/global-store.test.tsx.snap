// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GlobalStore onChange username 1`] = `
Array [
  <input
    class="input"
    name="username"
    placeholder="Username"
    role="textbox"
    value="test"
  />,
  <input
    class="input"
    name="profile.firstName"
    placeholder="First name"
    role="textbox"
    value=""
  />,
  <input
    class="input"
    name="profile.lastName"
    placeholder="Last name"
    role="textbox"
    value=""
  />,
]
`;

exports[`SimpleForm onChange username 1`] = `
Array [
  <input
    class="input"
    name="username"
    placeholder="Username"
    role="textbox"
    value="test"
  />,
  <input
    class="input"
    name="profile.firstName"
    placeholder="First name"
    role="textbox"
    value=""
  />,
  <input
    class="input"
    name="profile.lastName"
    placeholder="Last name"
    role="textbox"
    value=""
  />,
]
`;
